File: tttyoung/week_4_자료구조/boj-18258_큐2.py
import sys
from collections import deque #큐에서는 pop을 하면 맨 앞의 요소가 나오게 되고 나머지 요소들이 한칸씩 왼쪽으로 이동해야하므로 리스트를 사용하면 비효율적임
#deque를 사용하면 원래 큐를 사용할대의 시간복잡도인 O(n)에서 O(1)이 됨

num = int(sys.stdin.readline())
dq = deque()
for i in range(num):
    cmlistqu = sys.stdin.readline().split()

    if cmlistqu[0] == "push":
        dq.append(int(cmlistqu[1]))

    elif cmlistqu[0] == "pop":
        if len(dq) == 0:
            print(-1)
        else:
            print(dq.popleft()) #가장 왼쪽 요소가 나옴

    elif cmlistqu[0] == "size":
        print(len(dq))

    elif cmlistqu[0] == "empty":
        if len(dq) == 0:
            print(1)
        else:
            print(0)

    elif cmlistqu[0] == "front":
        if len(dq) == 0:
            print(-1)
        else:
            print(dq[0])

    elif cmlistqu[0] == "back":
        if len(dq) == 0:
            print(-1)
        else:
            print(dq[-1])

File: miggule2/week25/BOJ_18258_큐2.java
package week25;

import java.util.*;
import java.io.*;

public class BOJ_18258_큐2 {
    public static void main(String[] args)throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        int n = Integer.parseInt(br.readLine());

        LinkedList<Integer> queue = new LinkedList<>();
        int size = 0;

        for(int i = 0; i < n; i++){
            StringTokenizer st = new StringTokenizer(br.readLine());
            switch(st.nextToken()){
                case "push" :
                    int num = Integer.parseInt(st.nextToken());
                    queue.addFirst(num);
                    size++;
                    break;

                case "pop" :
                    if(size == 0){
                        bw.write(-1+"\n");
                        break;
                    }
                    size--;
                    bw.write(queue.removeLast()+"\n");
                    break;

                case "size" :
                    bw.write(size+"\n");
                    break;

                case "empty" :
                    if(size == 0) bw.write(1+"\n");
                    else bw.write(0+"\n");
                    break;

                case "front" :
                    if(size == 0) bw.write(-1+"\n");
                    else bw.write(queue.getLast()+"\n");
                    break;

                case "back" :
                    if(size == 0) bw.write(-1+"\n");
                    else bw.write(queue.getFirst()+"\n");
                    break;
                default : break;
            }
        }

        bw.flush();
        bw.close();
    }
}
